qiskit/circuit/quantumcircuit.py

```diff
@@ -1152,17 +1152,26 @@ def __init__(
         """The unit that :attr:`duration` is specified in."""
         self.metadata = {} if metadata is None else metadata
         """Arbitrary user-defined metadata for the circuit.
-
+ 
         Qiskit will not examine the content of this mapping, but it will pass it through the
         transpiler and reattach it to the output, so you can track your own metadata."""
 
     @classmethod
-    def _from_circuit_data(cls, data: CircuitData) -> typing.Self:
+    def _from_circuit_data(cls, data: CircuitData, add_regs: bool = False) -> typing.Self:
         """A private constructor from rust space circuit data."""
         out = QuantumCircuit()
-        out._data = data
         out._qubit_indices = {bit: BitLocations(index, []) for index, bit in enumerate(data.qubits)}
         out._clbit_indices = {bit: BitLocations(index, []) for index, bit in enumerate(data.clbits)}
+        out._data = data
+
+        # To add consistent registers, we compose the circuit onto a new one with registers,
+        # as each instruction must be mapped to qubits that belong to a register (in the circuit
+        # data coming from Rust, the qubits are not attached to any register).
+        if add_regs:
```

Comment:
Why not do something like:

```python
qr = QuantumRegister(bits=out.qubits)
out.add_register(qr)
```

Reply:
I tried, but then the qubits in the `CircuitData` are unaware that they should belong to `qr` and doing circuit comparisons still fails. For example if we had
```python
py_circ = QuantumCircuit(1)
py_circ.x(0)

rust_circ = QuantumCircuit._from_circuit_data(data, add_regs=True)  # data just contains the X gate on qubit 0

rust_circ == py_circ
```
would still evaluate to `False`.

Reply:
Would it be possible to compute `num_qubits` and `num_clbits` from `data`, and change the line `out = QuantumCircuit()` to `out = QuantumCircuit(num_qubits, num_clbits)`?

Reply:
The problem is that the instructions in `CircuitData` use qubits, that are not attached to any registers. If we do `out = QuantumCircuit(num_qubits, num_clbits)`, then we create a register for the circuit, but the qubits in the `CircuitData` do not belong to these (and we actually end up with twice the number of qubits) 🙂